### PR TITLE
Hardnet8

### PR DESCRIPTION
**Pull request by ducha-aiki**
### Description
Added HardNet8 -- new version of HardNet, considerably better than original one.

mAA 0.66046 vs 0.63196 for original at Image Matching Challenge 2020
https://www.cs.ubc.ca/research/image-matching-challenge/submissions/sid-00621-results/
### Status
**Ready**

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/feature/hardnet.py [hunk 1]
```diff
@@ -165,5 +166,5 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
         x_norm: torch.Tensor = self._normalize_input(input)
         x_features: torch.Tensor = self.features(x_norm)
         x_prePCA = F.normalize(x_features.view(x_features.size(0), -1))
-        pca = torch.mm(x_prePCA - self.mean, self.components.T)
+        pca = torch.mm(x_prePCA - self.mean, self.components)
```
Comment (edgarriba) [inline kornia/feature/hardnet.py]: @ducha-aiki can we have an inhouse pca near future ? :)

**Review by edgarriba — APPROVED** [review]
LGTM

File: kornia/feature/hardnet.py (hunk 1 above)
Comment (ducha-aiki) [reply kornia/feature/hardnet.py]: So far learning is done by sklearn, the only inference is torch. But - yes, why not?

File: test/feature/test_hardnet.py [hunk 2]
```diff
@@ -37,3 +37,34 @@ def test_jit(self, device, dtype):
...
+        hardnet = HardNet8().to(patches.device, patches.dtype)
+        assert gradcheck(hardnet, (patches,), eps=1e-4, atol=1e-4,
+                         raise_exception=True, )
+
+    @pytest.mark.jit
+    def test_jit(self, device, dtype):
+        B, C, H, W = 2, 1, 32, 32
+        patches = torch.ones(B, C, H, W, device=device, dtype=dtype)
+        model = HardNet8().to(patches.device, patches.dtype).eval()
+        model_jit = torch.jit.script(HardNet8().to(patches.device, patches.dtype).eval())
```
Comment (edgarriba) [inline test/feature/test_hardnet.py:69]: this one seems to fail
